File: src/cordova/Cordova/Cordova/CommandQueue.m
// Copyright (c) 2014 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "CommandQueue.h"

#import "CDVInvokedUrlCommand.h"
#import "CDVJSON.h"
#import "CDVPlugin.h"
#import "CrosswalkLite/Invocation.h"
#import "NSMutableArray+QueueAdditions.h"

@interface CommandQueue()

@property(nonatomic, strong) NSMutableArray* queue;

@end

@implementation CommandQueue

- (id)init
{
    self = [super init];
    if (self) {
        self.queue = [[NSMutableArray alloc] init];
    }
    return self;
}

- (BOOL)enqueueCommandBatch:(NSString*)batchJSON
{
    if ([batchJSON length] == 0) {
        return NO;
    }

    NSArray *commands = (NSArray*)[batchJSON JSONObject];
    for (NSInteger i = 0; i < commands.count; i++) {
        [self.queue addObject:[CDVInvokedUrlCommand commandFromJson:(NSArray*)commands[i]]];
    }
    return YES;
}

- (BOOL)execute:(CDVInvokedUrlCommand*)command;
{
    if (!command.className || !command.methodName) {
        NSLog(@"ERROR: Classname and/or methodName not found for command.");
        return NO;
    }

    CDVPlugin* plugin = [self.delegate getPluginInstance:command.className];
    if (!plugin) {
        NSLog(@"ERROR: Failed to find plugin instance for class: %@", command.className);
        return NO;
    }

    SEL sel = NSSelectorFromString([command.methodName stringByAppendingString:@":"]);
    if ([plugin respondsToSelector:sel]) {
        [Invocation call:plugin selector:sel arguments:[NSArray arrayWithObject:command]];
    } else {
        NSLog(@"ERROR: Method %@ not defined in plugin: %@", command.methodName, command.className);
        return NO;
    }

    return YES;
}

- (void)executePending
{
    while (self.queue.count) {
        CDVInvokedUrlCommand* command = [self.queue dequeue];
        if (![self execute:command]) {
            NSLog(@"ERROR: Failed to execute command");
        }
    }
}

@end

File: src/cordova/Cordova/Cordova/CommandQueue.swift
// Copyright (c) 2014 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation

protocol CommandQueueDelegate : class {
    func getPluginInstance(className: String) -> CDVPlugin?
}

class CommandQueue : NSObject {
    private var queue : Array<CDVInvokedUrlCommand> = []
    weak var delegate : CommandQueueDelegate?

    func enqueueCommandBatch(batchJSON: String) -> Bool {
        if batchJSON.lengthOfBytesUsingEncoding(NSUTF8StringEncoding) == 0 {
            return false
        }

        if let commands: NSArray = batchJSON.JSONObject() as? NSArray {
            for obj in commands {
                if let command: NSArray = obj as? NSArray {
                    queue.append(CDVInvokedUrlCommand(json: command))
                }
            }
            return true
        }
        return false
    }

    func execute(command: CDVInvokedUrlCommand) -> Bool {
        if command.className == nil || command.methodName == nil {
            return false
        }

        if let plugin: CDVPlugin = delegate?.getPluginInstance(command.className) {
            let selector: Selector = NSSelectorFromString("\(command.methodName):")
            if plugin.respondsToSelector(selector) {
                Invocation.call(plugin, selector: selector, arguments: [command])
                return true
            } else {
                println("ERROR: Method \(command.methodName) not defined in plugin: \(command.className)")
            }
        }
        return false
    }

    func executePending() {
        while queue.count > 0 {
            if !self.execute(queue.removeAtIndex(0)) {
                println("Failed to execute command")
            }
        }
    }
}
